### PR TITLE
Source file status dictates make rebuilds

### PR DESCRIPTION
**Pull request by lrvick**
## Scope

- [x] artifact targets now track their respective source directories for changes where possible
- [x] toolchain now provides: `toolchain-dist-cache` and `toolchain-restore-mtime`
  - Example: ```make toolchain-restore-mtime toolchain-dist-cache out/some-artifact```
    - sets all timestamps in repo to their git commit timestamps
    -  copies existing dist artifacts
    - NOOP unless  source files is newer for given artifact
    - can be used locally for dev or in CI to avoid wasted build time
- [x] ```make dist```
  - uses toolchain-restore-mtime and toolchain-dist-cache by default
  - now only builds artifacts if source files are newer than previously built artifacts
- [x] Release cut against main ready for review

## Review

- [ ] 1. fresh clone repo to this branch (with lfs and submodules)
- [ ] 2. review changes locally with ```git diff main```
- [ ] 3. ```make dist```
- [ ] 4. observe that only dist/release.env has changed, and it builds almost instantly
- [ ] 5. make a minor change to a file such as ```src/qos_enclave/src/main.rs```
- [ ] 6. run ```make```
- [ ] 7. confirm only out/qos_enclave.linux-x86_64 and out/qos_enclave.oci.x86_64.tar are rebuilt
- [ ] 8. `git restore` the modified file(s)
- [ ] 9. run ```make reproduce``` and ensure it is successful
- [ ] 10. Push signature with ```git sig add --push```
- [ ] 11. merge branch to main

### CODE REVIEW

File: .github/workflows/artifacts.yml [hunk 1]
```diff
@@ -142,13 +142,18 @@ jobs:
...
         with:
           fail-on-cache-miss: true
-          key: toolchain-rust-${{ hashFiles('Makefile,config/make.env') }}
-          path: fetch/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot
+          key: toolchain-rust-${{ hashFiles('Makefile','config/make.env') }}
+          path: cache/aws/x86_64/src/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot
+
+      - name: Ensure rust cache is ignored by make
+        run: |
+          touch --date=@0 cache/aws/x86_64/src/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot
```
Comment (james-callahan) [inline .github/workflows/artifacts.yml]: Is this one really required?
I thought only the below one would be

File: Makefile [hunk 2]
```diff
@@ -14,7 +14,8 @@ KEYS := \
 .DEFAULT_GOAL :=
 .PHONY: default
 default: \
-	toolchain \
+	restore-mtime \
```
Comment (james-callahan) [inline Makefile]: I don't think order works here: make is free to build the targets in any order (or even in parallel if e.g. `-j 4` is passed)

File: Makefile [hunk 3]
```diff
@@ -14,7 +14,8 @@ KEYS := \
...
+	dist-cache \
```
Comment (james-callahan) [inline Makefile]: Same with this: this may not get ran first

File: Makefile (hunk 2 above)
Comment (lrvick) [reply Makefile]: I could add .NOTPARALELL to default

Comment (lrvick) [reply Makefile]: Okay .WAIT is what i want. Added that after restore-mtime and dist-cache

Comment (lrvick) [reply Makefile]: okay, abstracted this into toolchain with "toolchain-" prefixes so it is obvious what is going on.

**Pull request merged**